onmt/data/SampledDataset.lua

```diff
@@ -0,0 +1,266 @@
+end
+
+--[[ initiate sampling ]]
+function SampledDataset:sample(avgPpl)
+  _G.logger:info('Sampling...')
+
+  -- populate self.samplingProb with self.ppl if average ppl is below self.sample_w_ppl_init
+  if avgPpl == nil and self.sample_w_ppl then
+    avgPpl = torch.sum(self.ppl)
+    avgPpl = avgPpl/self.ppl:size(1)
+  end
+
+  if self.sample_w_ppl and not self.startedPplSampling and avgPpl < self.sample_w_ppl_init then
+    _G.logger:info('Beginning to sample with ppl as probability distribution...')
+    self.startedPplSampling = true
+  end
+
+  if self.startedPplSampling then
+
+    local threshold = self.sample_w_ppl_max
+
+    if self.sample_w_ppl_max < 0 then
+      -- use mode (instead of mean) and stdev of samples with ppl>=mode to
+      -- find max ppl to consider (mode + x * stdev). when x is:
+      --      x: 1 ~ 100% - 31.7%/2 of train data are not included (divide by 2 because we cut only one-tail)
+      --      x: 2 ~ 100% - 4.55%/2
+      --      x: 3 ~ 100% - 0.270%/2
+      --      x: 4 ~ 100% - 0.00633%/2
+      --      x: 5 ~ 100% - 0.0000573%/2
+      --      x: 6 ~ 100% - 0.000000197%/2
+      --  (https://en.wikipedia.org/wiki/Standard_deviation)
+      -- we are using mode instead of average, and only samples above mode to calculate stdev, so
+      -- this is not really theoretically valid numbers, but more for emperical uses
+
+      local x = math.abs(self.sample_w_ppl_max)
+
+      -- find mode
+      local pplRounded = torch.round(self.ppl)
+      local bin = {}
+      for i = 1, pplRounded:size(1) do
+        if self.ppl[i] ~=  self.sample_w_ppl_init then
+          local idx = pplRounded[i]
+          if bin[idx] == nil then
+            bin[idx] = 0
+          end
+          bin[idx] = bin[idx] + 1
+        end
+      end
+      local mode = nil
+      for key,value in pairs(bin) do
+        if mode == nil or bin[mode] < value then
+          mode = key
+        end
+      end
+
+      -- stdev with mode only using samples with ppl >= mode
+      local sum = 0
+      local cnt = 0
+      for i = 1, self.ppl:size(1) do
+        if self.ppl[i] > mode and self.ppl[i] ~= self.sample_w_ppl_init then
+          sum = math.pow(self.ppl[i]-mode, 2)
+          cnt = cnt + 1
+        end
+      end
+      local stdev = math.sqrt(sum/(cnt-1))
+
+      threshold = mode + x * stdev
+
+      _G.logger:info('Sampler count: ' .. cnt)
+      _G.logger:info('Sampler mode: ' .. mode)
+      _G.logger:info('Sampler stdev: ' .. stdev)
+      _G.logger:info('Sampler threshold: ' .. threshold)
+    end
+
+    for i = 1, self.ppl:size(1) do
+      if self.ppl[i] ~= self.sample_w_ppl_init and self.ppl[i] > threshold then
+        -- asign low value to instances with ppl above threshold (outliers)
+        self.samplingProb[i] = 1
+      else
+        self.samplingProb[i] = self.ppl[i]
+      end
+    end
+  end
+
+--  local sampled = torch.multinomial(self.samplingProb:view(1,#self.src), self.samplingSize, --[[replacement]] false)
```

Comment:
I'd like to keep this comment even if it is dead code just as a reminder of how it is done by generic torch library. Maybe the torch function will be updated for speed improvement in the future.